### PR TITLE
Better stream eventing

### LINKED ISSUE #360: handle changeListener inputs in transactions correctly
This is something that would help us in handling the changes made by transactions and lessen the amount of updates that take place.

Currently, database changes made in `@transaction` functions immediately notify the changeListener and start to update the table in streams. This has the following problems:
- The tables are updated, even if the transaction is not finished yet (and therefore no changes have been persisted yet)
- If a transaction e.g. deletes some entries in a table and immediately inserts them again, two notifications are put inside the changeListener, adding unnecessary refresh cycles. 
- When a stream (e.g. on joins) has to listen on multiple entity names at once, even more unnecessary refreshes happen.
- If the transaction fails, the streams will get updated even if nothing happened.

Because of the reasons above, I propose to change the type of the changeListener from `Stream<String>` to `Stream<Set<String>>` and put a proxy `changeListener` into the transaction generation (Line 32, see below), which buffers all changes and then submits them together once the transaction was completed successfully. This adds some overhead into transactions (for the proxy) and some smaller overhead for streams, which now have calculate an intersection, instead of a simple contains calculation for filtering relevant updates.

I want to hear your input first, @vitusortner and I think that I probably won't start to work on this soon. (I'm currently having more fun with the sqlparser :wink: )

https://github.com/vitusortner/floor/blob/18117ed397b5787905576aedad90e0f347931792/floor_generator/lib/writer/transaction_method_writer.dart#L27-L36

A small note: I was surprised to see that adding this to transactions is far easier than I imagined.

### PR DESCRIPTION
**Pull request by mqus**
This changes the datatype of the events for streaming to `Set<String>` and implements two features that benefit from that:
- If a delete/update also changes other tables via foreign-key relationships, then streams on those entities also get updated, depending on the columns and types of relationships, even transitively.
- Changes made in a transaction will trigger all affected tables at the end (and not before), and only if the transaction succeeded.

What did not change: the handling of streams of views. They will still get updated by *any* changes in the database, since those dependencies can only be resolved with a query parser/analyzer.

To summarize, this will cause better stream updates for entities with FK-relationships and transaction events, and while also outputting less events than before.

It is probably best to review this commit-by-commit.

Addresses #373 partially.
Closes #360

### CODE REVIEW

**Comment by mqus** [comment]
I just noticed that I missed `@insert`s `onConflict: replace` and similar options, which also cause cascading changes, I'll have to consider this.

**Review by vitusortner — COMMENTED** [review]
I'll continue the review ASAP. So far, I left some questions.

File: floor_generator/lib/misc/foreign_key_map.dart [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import 'package:floor_annotation/floor_annotation.dart' show ForeignKeyAction;
+import 'package:floor_generator/value_object/entity.dart';
+import 'package:floor_generator/value_object/foreign_key.dart';
+import 'package:floor_generator/misc/extension/foreign_key_action_extension.dart';
+
+class ForeignKeyMap {
+  final Map<String, Map<ForeignKey, Entity>> fkDependencies;
+
+  ForeignKeyMap.fromEntities(Iterable<Entity> stuff)
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor_generator/lib/misc/foreign_key_map.dart:9]: whoops :D

File: floor/lib/src/adapter/deletion_adapter.dart [hunk 2]
```diff
@@ -8,21 +8,21 @@ class DeletionAdapter<T> {
...
-    final StreamController<String>? changeListener,
+    final void Function()? changeHandler,
   ])  : assert(entityName.isNotEmpty),
         assert(primaryKeyColumnName.isNotEmpty),
         _database = database,
         _entityName = entityName,
         _primaryKeyColumnNames = primaryKeyColumnName,
         _valueMapper = valueMapper,
-        _changeListener = changeListener;
+        _changeHandler = changeHandler ?? (() {/* do nothing */});
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor/lib/src/adapter/deletion_adapter.dart:25]: Checking for null on each invocation might incur some very minor runtime cost but apart from that I have no real reason. If you prefer this version then I'll change it.

File: floor/lib/src/adapter/insertion_adapter.dart [hunk 3]
```diff
@@ -42,7 +42,7 @@ class InsertionAdapter<T> {
       );
     }
     await batch.commit(noResult: true);
-    _changeListener?.add(_entityName);
+    _changeListener?.add({_entityName});
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor/lib/src/adapter/insertion_adapter.dart:45]: Actually, the insertion_adapter also needs lambdas because of the onconflict:update issue... I thought I committed that already but it seems like I didn't...

[the remark this answers is not in the document]
Comment (mqus) [reply floor/lib/src/adapter/insertion_adapter.dart:45]: ah right, I didn't do this because the `onConflict` clause can be different for each call of `insertX`, altering the entities which have to be notified... And I could not decide what is best.

There are basically three main options here:
1. always assume `onConflict:Update` if it occurs even once for a `InsertionAdapter.insert` call
2. give the `InsertionAdapter` two different sets/lambdas via the constructor and it will choose which one to use for triggering updates, based on the `onConflict` clause given in the `InsertionAdapter.insert` call
3. insert the lambda/set via the invocation of the `InsertionAdapter.insert` call and not in the constructor

I'm not sure which is best here, but 1 seems to be suboptimal to me.

**Comment by saharvx9** [comment]
following working awsome!
